koku/subs/subs_data_extractor.py

```diff
@@ -0,0 +1,170 @@
+        """Determine the latest processed timestamp for a provider for a given month and year."""
+        with schema_context(self.schema):
+            last_time = SubsLastProcessed.objects.filter(
+                source_uuid=self.provider_uuid, year=year, month=month
+            ).first()
+        return last_time.latest_processed_time if last_time else None
+
+    def determine_end_time(self, year, month):
+        sql = (
+            f" SELECT MAX(lineitem_usagestartdate) FROM aws_line_items"
+            f" WHERE source='{self.provider_uuid}' AND year='{year}' AND month='{month}'"
+        )
+        latest = self._execute_trino_raw_sql_query(sql, log_ref="insert_subs_last_processed_time")
+        return latest[0][0]
+
+    def determine_line_item_count(self, where_clause):
+        """Determine the number of records in the table that have not been processed and match the criteria"""
+        table_count_sql = f"SELECT count(*) FROM {self.schema}.{self.table} {where_clause}"
+        count = self._execute_trino_raw_sql_query(table_count_sql, log_ref="determine_subs_processing_count")
+        return count[0][0]
+
+    def determine_where_clause(self, latest_processed_time, end_time, year, month):
+        """Determine the where clause to use when processing subs data"""
+        return (
+            f"WHERE source='{self.provider_uuid}' AND year='{year}' AND month='{month}' AND"
+            " lineitem_productcode = 'AmazonEC2' AND lineitem_lineitemtype IN ('Usage', 'SavingsPlanCoveredUsage') AND"
+            " product_vcpu IS NOT NULL AND strpos(resourcetags, 'com_redhat_rhel') > 0 AND"
+            f" lineitem_usagestartdate > TIMESTAMP '{latest_processed_time}' AND"
+            f" lineitem_usagestartdate <= TIMESTAMP '{end_time}'"
+        )
+
+    def update_latest_processed_time(self, year, month, end_time):
+        """Update the latest processing time for a provider"""
+        with schema_context(self.schema):
+            subs_obj = SubsLastProcessed.objects.filter(
+                source_uuid=self.provider_uuid, year=year, month=month
+            ).first() or SubsLastProcessed.objects.create(
+                source_uuid=TenantAPIProvider.objects.get(uuid=self.provider_uuid), year=year, month=month
+            )
+            subs_obj.latest_processed_time = end_time
+            subs_obj.save()
+
+    def extract_data_to_s3(self, start_date, end_date):
+        """Process new subs related line items from reports to S3."""
+        LOG.info(log_json(self.tracing_id, msg="beginning subs rhel extraction", context=self.context))
+        month = start_date.strftime("%m")
+        year = start_date.strftime("%Y")
+        # if there is no latest time for this month, we need to include the first hour of the month
+        latest_timestamp = self.determine_latest_processed_time_for_provider(year, month) or (
+            datetime(start_date.year, start_date.month, start_date.day) - timedelta(hours=1)
+        )
```

Comment:
I was operating under the assumption that if there was not a `last_processed_time` it was due to initial ingestion, that was a poor assumption and I have adjusted the date logic to always pass in the month start because if there is not a latest processed time for the month, we want to process the entire month. 

The reason for the `- timedelta(hours=1)` is the query is just GT not GTE so we don't have to manipulate the latest timestamp, so we go back one hour to the previous month as the timestamp so the 0th hour of the current month is checked. I have updated the comment here as well to be more descriptive.